pkg/policyviolation/namespacedpv.go

```diff
@@ -0,0 +1,161 @@
+	kyvernov1alpha1 "github.com/nirmata/kyverno/pkg/client/clientset/versioned/typed/kyverno/v1alpha1"
+	kyvernolister "github.com/nirmata/kyverno/pkg/client/listers/kyverno/v1alpha1"
+	dclient "github.com/nirmata/kyverno/pkg/dclient"
+	engine "github.com/nirmata/kyverno/pkg/engine"
+	labels "k8s.io/apimachinery/pkg/labels"
+)
+
+func CreateNamespacePV(pvLister kyvernolister.NamespacedPolicyViolationLister, client *kyvernoclient.Clientset, engineResponses []engine.EngineResponse) {
+	// var pvs []kyverno.NamespacedPolicyViolation
+	// for _, er := range engineResponses {
+	// 	// ignore creation of PV for resoruces that are yet to be assigned a name
+	// 	if er.PolicyResponse.Resource.Name == "" {
+	// 		glog.V(4).Infof("resource %v, has not been assigned a name, not creating a namespace policy violation for it", er.PolicyResponse.Resource)
+	// 		continue
+	// 	}
+
+	// 	if !er.IsSuccesful() {
+	// 		glog.V(4).Infof("Building namespace policy violation for engine response %v", er)
+	// 		if pv := buildNamespacedPVForPolicy(er); !reflect.DeepEqual(pv, kyverno.NamespacedPolicyViolation{}) {
+	// 			pvs = append(pvs, pv)
+	// 		}
+	// 	}
+	// }
+
+	// createNamespacedPV(pvLister, client, pvs)
+}
+
+// CreateNamespacedPVWhenBlocked creates pv on resource owner only when admission request is denied
+func CreateNamespacedPVWhenBlocked(pvLister kyvernolister.NamespacedPolicyViolationLister, client *kyvernoclient.Clientset,
+	dclient *dclient.Client, engineResponses []engine.EngineResponse) {
+	// var pvs []kyverno.NamespacedPolicyViolation
+	// for _, er := range engineResponses {
+	// 	// child resource is not created in this case thus it won't have a name
+	// 	glog.V(4).Infof("Building policy violation for denied admission request, engineResponse: %v", er)
+	// 	if pvList := buildNamespacedPVWithOwner(dclient, er); len(pvList) != 0 {
+	// 		pvs = append(pvs, pvList...)
+	// 		glog.V(3).Infof("Built policy violation for denied admission request %s/%s/%s",
+	// 			er.PatchedResource.GetKind(), er.PatchedResource.GetNamespace(), er.PatchedResource.GetName())
+	// 	}
+	// }
+	// createNamespacedPV(pvLister, client, pvs)
+}
+
+func buildNamespacedPV(info Info) kyverno.NamespacedPolicyViolation {
+	return buildNamespacedPVObj(info.PolicyName,
+		kyverno.ResourceSpec{
+			Kind:      info.Resource.GetKind(),
+			Namespace: info.Resource.GetNamespace(),
+			Name:      info.Resource.GetName(),
+		},
+		info.Rules)
+}
+
+//buildNamespacedPVObj returns an value of type PolicyViolation
+func buildNamespacedPVObj(policy string, resource kyverno.ResourceSpec, fRules []kyverno.ViolatedRule) kyverno.NamespacedPolicyViolation {
+	pv := kyverno.NamespacedPolicyViolation{
+		Spec: kyverno.PolicyViolationSpec{
+			Policy:        policy,
+			ResourceSpec:  resource,
+			ViolatedRules: fRules,
+		},
+	}
+
+	labelMap := map[string]string{
+		"policy":   policy,
+		"resource": strings.Join([]string{resource.Kind, resource.Namespace, resource.Name}, "."),
```

Comment:
Cluster policy violation has labels patched by policy controller, while namespaced policy violation adds labels while creating the resource. Would be nice to be consistent.

Reply:
Added when building cluster policy violation as it avoids update on the object.

Reply:
ok, as the policy controller has the logic in resource manager to claim and release namespaced violations where it adds and removed ownerReferences, so thats y suggested before using the approach of cluster policy to claim a violation.

But with setting the labels while creating resource, we don't really need the claim and release logic.